### PR TITLE
Flink: add 1.20 support and remove 1.17

### PR DESCRIPTION
**Pull request by stevenzwu**
The 4th commit has the real change. also fixed jmh setup issue in 1.19 before.

### CODE REVIEW

File: flink/v1.19/flink/src/jmh/java/org/apache/iceberg/flink/sink/shuffle/MapRangePartitionerBenchmark.java [hunk 1]
```diff
@@ -66,6 +69,8 @@ public class MapRangePartitionerBenchmark {
           Types.NestedField.required(9, "name9", Types.StringType.get()));
 
   private static final SortOrder SORT_ORDER = SortOrder.builderFor(SCHEMA).asc("id").build();
+  private static final Comparator<StructLike> SORT_ORDER_COMPARTOR =
```
Comment (stevenzwu) [inline flink/v1.19/flink/src/jmh/java/org/apache/iceberg/flink/sink/shuffle/MapRangePartitionerBenchmark.java:72]: there was jmh setup issue (not included) in 1.19. hence the code issue wasn't caught earlier.

File: flink/v1.20/flink/src/test/java/org/apache/iceberg/flink/source/TestIcebergSpeculativeExecutionSupport.java [hunk 2]
```diff
@@ -165,7 +165,7 @@ private static Configuration configure() {
     Configuration configuration = new Configuration();
     configuration.set(CoreOptions.CHECK_LEAKED_CLASSLOADER, false);
     configuration.set(RestOptions.BIND_PORT, "0");
-    configuration.set(JobManagerOptions.SLOT_REQUEST_TIMEOUT, 5000L);
+    configuration.set(JobManagerOptions.SLOT_REQUEST_TIMEOUT, Duration.ofSeconds(5));
```
Comment (stevenzwu) [inline flink/v1.20/flink/src/test/java/org/apache/iceberg/flink/source/TestIcebergSpeculativeExecutionSupport.java:168]: Flink changed the config option type from long to Duration

**Comment by openinx** [comment]
Hi @stevenzwu   Is it possible to keep the original commits and authors in the newly introduced flink1.20 , so that people could easily to track how was the current version iterated ?

**Comment by stevenzwu** [comment]
> Hi @stevenzwu Is it possible to keep the original commits and authors in the newly introduced flink1.20 , so that people could easily to track how was the current version iterated ?

@openinx yes, we don't know `squash and merge` for this type of PR and keep individual commits. The first commit is from `git mv flink/v1.19 flink/v1.20`, which will carry over the history where the file is renamed from.

e.g. https://github.com/apache/iceberg/commits/main/flink/v1.19/build.gradle
image

File: flink/v1.20/flink/src/jmh/java/org/apache/iceberg/flink/sink/shuffle/MapRangePartitionerBenchmark.java [hunk 3]
```diff
@@ -140,8 +139,7 @@ private static String randomString(String prefix) {
       buffer[i] = (byte) CHARS.charAt(ThreadLocalRandom.current().nextInt(CHARS.length()));
     }
 
-    // CHARS is all ASCII
-    return prefix + new String(buffer, StandardCharsets.US_ASCII);
+    return prefix + new String(buffer);
```
Comment (pvary) [inline flink/v1.20/flink/src/jmh/java/org/apache/iceberg/flink/sink/shuffle/MapRangePartitionerBenchmark.java:142]: Do we need to add this change to the 1.19 test as well?

File: flink/v1.20/flink/src/main/java/org/apache/iceberg/flink/FlinkCatalog.java [hunk 4]
```diff
@@ -401,7 +401,6 @@ void createIcebergTable(ObjectPath tablePath, ResolvedCatalogTable table, boolea
 
     Schema icebergSchema = FlinkSchemaUtil.convert(table.getResolvedSchema());
     PartitionSpec spec = toPartitionSpec(((CatalogTable) table).getPartitionKeys(), icebergSchema);
-
```
Comment (pvary) [inline flink/v1.20/flink/src/main/java/org/apache/iceberg/flink/FlinkCatalog.java:404]: What is this change?

**Review by pvary — APPROVED** [review]

**Comment by stevenzwu** [comment]
thanks @pvary for the review. Did `rebase and merge` to preserve the Individual commits

**Pull request merged**